cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java

```diff
@@ -0,0 +1,263 @@
+    prevMinute = new Date(System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(1));
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(prevMinute);
+    cal.set(Calendar.SECOND, 0);
+    cal.set(Calendar.MILLISECOND, 0);
+    prevMinute = cal.getTime();
+    currentTime = new SimpleDateFormat("yyyy-MM-dd-HH-mm").format(prevMinute);
+
+    Job job = context.getHadoopJob();
+    job.getConfiguration().set("fs.s3n.awsAccessKeyId", config.accessID);
+    job.getConfiguration().set("fs.s3n.awsSecretAccessKey", config.accessKey);
+    if (config.fileRegex != null) {
+      job.getConfiguration().set("input.path.regex", config.fileRegex);
+    }
+    job.getConfiguration().set("input.path.name", config.path);
+    if (config.timeTable != null) {
+      table = context.getDataset(config.timeTable);
+    }
+
+    FileInputFormat.setInputPathFilter(job, S3Filter.class);
+    FileInputFormat.addInputPath(job, new Path(config.path));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Text> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_SCHEMA)
+      .set("ts", System.currentTimeMillis())
+      .set("body", input.getValue().toString())
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    if (succeeded && table != null && "timefilter".equals(config.fileRegex)) {
+      table.write("lastTimeRead", currentTime);
+    }
+  }
+
+  /**
+   * Filter class to filter out filenames in the input path.
+   */
+  public static class S3Filter extends Configured implements PathFilter {
+    private boolean useTimeFilter;
+    private Pattern regex;
+    private String pathName;
+
+    @Override
+    public boolean accept(Path path) {
+      String filename = path.toString();
+      //InputPathFilter will first check the directory if a directory is given
+      if (filename.equals(pathName) || filename.equals(pathName + "/")) {
+        return true;
+      }
+
+      if (useTimeFilter) {
+        //use stateful time filter
+        if (table != null) {
+          String lastRead = Bytes.toString(table.read("lastTimeRead"));
+          SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd-HH-mm");
+          Date dateLastRead;
+          try {
+            dateLastRead = sdf.parse(lastRead);
+          } catch (Exception e) {
+            dateLastRead = new Date();
+            dateLastRead.setTime(0);
+          }
+
+          Date fileDate;
+          try {
+            fileDate = sdf.parse(path.getName().substring(0, DATE_LENGTH));
+          } catch (ParseException pe) {
+            //this should never happen
+            LOG.warn("Couldn't parse file: " + path.getName());
+            fileDate = prevMinute;
+          }
+
+          if (fileDate.compareTo(dateLastRead) > 0 && fileDate.compareTo(prevMinute) <= 0) {
```

Comment:
Also, should this be replaced by
```java
return fileDate.compareTo(dateLastRead) > 0 && fileDate.compareTo(prevMinute) <= 0;
```
or do you think that this is better for clarity?